### DISCUSSION TITLE
Problem: Handling Deep Cloning of Objects in JavaScript

### DISCUSSION THREAD

**Discussion by vladimir-de-oppenheimer** [community/community#139491] [Programming Help]
### Body

I’m working with complex objects in JavaScript and need to create deep copies of these objects. I tried using `Object.assign()` and the spread operator (`...`), but these methods only create shallow copies. As a result, nested objects or arrays within the object still reference the same memory, causing unexpected behavior when modifying the cloned object.

### Example Code:

```javascript
const originalObject = {
  name: "John",
  address: {
    street: "123 Main St",
    city: "New York"
  }
};

const shallowCopy = { ...originalObject };

shallowCopy.address.city = "San Francisco";

console.log(originalObject.address.city); // Outputs: "San Francisco"

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by sabbirsami** [comment]
# Solution: Deep Cloning Objects in JavaScript

## Problem Overview

When working with complex objects in JavaScript, methods like `Object.assign()` and the spread operator (`...`) only create shallow copies. This means that nested objects or arrays within the copied object still reference the same memory. To avoid this, we need to perform a **deep clone**, where every level of the object is duplicated independently.

## Solution

There are several ways to perform deep cloning in JavaScript, each with its own advantages and limitations.

### 1. **Using `JSON.parse()` and `JSON.stringify()`**

This method is simple and works well for most cases, but it has limitations (e.g., it won’t clone functions or handle `undefined` values).

```javascript
const originalObject = {
  name: "John",
  address: {
    street: "123 Main St",
    city: "New York"
  }
};

const deepCopy = JSON.parse(JSON.stringify(originalObject));

deepCopy.address.city = "San Francisco";

console.log(originalObject.address.city); // Outputs: "New York"

**Reply by ebndev** [reply]
Hi @sabbirsami, thanks for participating in GitHub Discussions!

Please note: we’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  

You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. 

Please feel encouraged to continue engaging in Discussions as your authentic self though. We appreciate you being a part of the Community 🌟

**Comment by ebndev** [comment]
Hi @vladimir-de-oppenheimer & @sabbirsami, thanks for being a part of the GitHub Community!

While we appreciate you wanting to contribute to Discussions, please note: We made the decision to disable the ability to earn Achievements in our Community in order to discourage users from participating in coordinated or inauthentic activity like rapid questions and answers in order to earn badges. You can learn more about this decision in our announcement post here [Achievements will no longer be available in the Community](https://github.com/orgs/community/discussions/106536).

Note that GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this, and other related posts. 

Any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
